src/LowLevelMotionController.cpp

```diff
@@ -231,36 +265,120 @@ int main(int argc, char **argv)
+                }
+                else
+                {
+                    ROS_ERROR("Unknown ground interaction type");
+                    server.setAborted();
+                }
+            }
+
             //  This will contain the target twist or velocity that we want to achieve
             geometry_msgs::TwistStamped target_twist;
+            iarc7_msgs::OrientationThrottleStamped uav_command;
 
             // Check for a safety state in which case we should execute our safety response
             if(safety_client.isSafetyActive())
             {
                 // This is the safety response
-                // Override whatever the Acceleration Planner wants to do and attempt to land
-                target_twist.twist.linear.z = -0.2; // Try to descend
+                bool success = landPlanner.resetForTakeover(current_time);
+                ROS_ASSERT_MSG(success, "LowLevelMotion LandPlanner resetForTakeover failed");
+                ROS_WARN("Transitioning to state LAND for safety response");
+                motion_state = MotionState::LAND;
             }
-            else
+            else if(motion_state == MotionState::VELOCITY_CONTROL)
             {
                 // If nothing is wrong get a target velocity from the acceleration planner
                 accelerationPlanner.getTargetTwist(current_time, target_twist);
+
+                // Request the appropriate throttle and angle settings for the desired velocity
+                quadController.setTargetVelocity(target_twist.twist);
+
+                // Get the next uav command that is appropriate for the desired velocity
+                bool success = quadController.update(current_time, uav_command);
+                ROS_ASSERT_MSG(success, "LowLevelMotion quad velocity controller update failed");
             }
+            else if(motion_state == MotionState::TAKEOFF)
+            {
+                bool success = takeoffController.update(current_time, uav_command);
+                ROS_ASSERT_MSG(success, "LowLevelMotion takeoff controller update failed");
+
+                if(takeoffController.isDone())
+                {
+                    ROS_DEBUG("Takeoff completed. Hover throttle: %f", takeoffController.getHoverThrottle());
+                    //thrust_model.something = takeoffController.getHoverThrottle();
+                    server.setSucceeded();
+
+                    success = quadController.resetForTakeover();
+                    quadController.setThrustModel(thrust_model);
+                    ROS_ASSERT_MSG(success, "LowLevelMotion switching to velocity control failed");
+                    motion_state = MotionState::VELOCITY_CONTROL;
+                }
+            }
+            else if(motion_state == MotionState::LAND) {
 
-            // Publish the current target velocity
-            uav_velocity_target_.publish(target_twist);
+                bool success = landPlanner.getTargetTwist(current_time, target_twist);
```

Comment:
I agree about the land planner providing the entire twist, but I don't think we need multiple planners, I would rather let the single planner switch frames.